### PR TITLE
[RNMobile] RangeCell styling refactor

### LINKED ISSUE #1448: BottomSheet - change style of RangeCell component
According to discussion [here](https://github.com/WordPress/gutenberg/pull/17282) we need to adjust Slider style before shipping it.

### PR DESCRIPTION
**Pull request by lukewalczak**
## Description

**NOTE**: That PR is based on `rnmobile/spacer`
Fixes: https://github.com/wordpress-mobile/gutenberg-mobile/issues/1448
Ref to gb-mobile: https://github.com/wordpress-mobile/gutenberg-mobile/pull/1511
Change `RangeCell` mobile component styling.

## How has this been tested?

1. Open ios or android app
2. Add `Spacer` component
3. Press settings button
4. Check `RangeCell` styling
## Screenshots 
ios dark mode unfocused | ios light mode unfocused
--- | ---
Screenshot 2019-10-29 at 10 39 29 | Screenshot 2019-10-29 at 10 41 09

ios dark mode focused | ios light mode focused
--- | ---
Screenshot 2019-10-29 at 11 12 03 | Screenshot 2019-10-29 at 10 41 14

android focused | android unfocused
--- | ---
Screenshot 2019-10-29 at 10 44 07 | Screenshot 2019-10-29 at 10 44 02

* Reset Button functionality

## Types of changes

* add reset button
* move slider to new line
* change input styling

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate.

### CODE REVIEW

**Comment by lukewalczak** [comment]
@iamthomasbishop please look at it 🙏

**Comment by iamthomasbishop** [comment]
@lukewalczak Looking good! I have just one thing that I think we might want to address before shipping, and two suggestions which aren't blockers.

* In Dark Mode, can we change the `Reset` button text and the "filled" part of the slider (everything left of the drag handle) to both use [blue 30](https://color-studio.blog) aka `#5198d9`?

Not blockers:

* Is there any way we can move the relative position of the Slider track row up slightly (8 or 9px if my measurements are correct) without messing up other components? I think every Slider instance will have the same. (This isn't a blocker, but would be nice to tighten up spacing.) Visual example:

Screen Shot 2019-10-29 at 9 40 33 AM

* Suggestion: we can probably remove the icon on the setting row

**Comment by lukewalczak** [comment]
Updated according to the comments above:

Screenshot 2019-10-30 at 13 27 58

**Comment by iamthomasbishop** [comment]
@lukewalczak Colors looks better, thank you! One thing I just realized: my last comment gave the incorrect value here –

> Is there any way we can move the relative position of the Slider track row up slightly (8 or 9px if my measurements are correct)

It should have been **12px**, not 8-9. The screenshot I provided was accurate. Apologies for the mix-up.

**Comment by lukewalczak** [comment]
Updated one more time: 
Screenshot 2019-10-30 at 16 41 06

Also we've discovered that slider on android has additional space before and after slider. I found an [issue](https://github.com/react-native-community/react-native-slider/issues/98) for that in community library. Don't know if this is a system behaviour, but proposed workaround within an issue is not satisfying (it requires more changes in our case).

**Comment by iamthomasbishop** [comment]
Changes are looking good, and if we need to address the extra spacing separately, that's fine. You've got my blessing to move forward and ship this 👍

File: packages/components/src/mobile/bottom-sheet/cell.native.js
Comment (pinarol) [inline packages/components/src/mobile/bottom-sheet/cell.native.js]: __ and _x functions are not working with variables, you need to pass static text in them.

Also do we need sprintf here?

File: packages/components/src/mobile/bottom-sheet/cell.native.js
Comment (pinarol) [inline packages/components/src/mobile/bottom-sheet/cell.native.js]: We need to localize these texts here

**Comment by pinarol** [comment]
I've changed the base branch since spacer is merged, but it seems like we have merge conflicts so I added In Progress label. Please remove the label after solving them.

**Comment by pinarol** [comment]
Also, is there a gutenberg-mobile PR? I couldn't find the link.

File: packages/components/src/mobile/bottom-sheet/cell.native.js
Comment (lukewalczak) [reply packages/components/src/mobile/bottom-sheet/cell.native.js]: Fixed

File: packages/components/src/mobile/bottom-sheet/cell.native.js
Comment (lukewalczak) [reply packages/components/src/mobile/bottom-sheet/cell.native.js]: Fixed

**Comment by lukewalczak** [comment]
Updated description with a link to gb-mobile PR and resolved conflicts.

**Comment by jbinda** [comment]
I have also resolve conflicts after merge [RangeControl markup](https://github.com/WordPress/gutenberg/pull/18036)

File: packages/components/src/mobile/bottom-sheet/range-cell.native.js
Comment (etoledom) [inline packages/components/src/mobile/bottom-sheet/range-cell.native.js]: It's interesting that the root `Cell` component provides `Reset` option, but the reset functionality has to be handled by the parent component. It feels to me more like a generic button that could be anything. Maybe we can add it as an accessory view on this component, similar how we added the Switch?

Just by curiosity, I added `allowReset` to other controls, it shows the Reset button but it looks a bit off (apart than it does nothing). It doesn't feel to me that it deserves to be on the root Cell component, unless we plan to add reset to more controls? (In which case this might be a good-enough first step). cc @iamthomasbishop 

IMG_3459

**Review by etoledom — COMMENTED** [review]
Thank you for these improvements @lukewalczak - they look great! 🎉 

Added a couple of questions (rather than requests) on the code comments.

And I noticed a few small detail testing with bigger text sizes:

- If the system font sizes are set to a bigger size (quite common size), the fields start cropping the last digit for numbers bigger than 1000. This can be super confusing.
- If the text size is set to the accessibility sizes (the super big ones), the layout start to break. This case is a bit more extreme and we don't really have it well covered in the library in general, so maybe it's not that urgent. But we would want in this case the reset button to be vertically aligned.

range

File: packages/components/src/mobile/bottom-sheet/range-cell.native.js
Comment (iamthomasbishop) [reply packages/components/src/mobile/bottom-sheet/range-cell.native.js]: > It doesn't feel to me that it deserves to be on the root Cell component, unless we plan to add reset to more controls? (In which case this might be a good-enough first step)

That's correct, I don't think there are any other planned components at the moment that will need an explicit `reset` button like this.

File: packages/components/src/mobile/bottom-sheet/cell.native.js
Comment (jbinda) [reply packages/components/src/mobile/bottom-sheet/cell.native.js]: you are right `allowReset` on Web is a `boolean` so it should be consistent on mobile. We take care of it

File: packages/components/src/mobile/bottom-sheet/range-cell.native.js
Comment (iamthomasbishop) [reply packages/components/src/mobile/bottom-sheet/range-cell.native.js]: One more note: if we do end up adding `reset` to other cells/components, we'll want to display them in the same two-row way which provides more space for the `reset` button.

Comment (jbinda) [reply packages/components/src/mobile/bottom-sheet/range-cell.native.js]: Currently we don't have many `settings` in block but maybe it's not bad idea to allow reset single field. Correct me if I'm wrong but it seems now it's possible to manually delete/change the value or clear all settings.

**Comment by jbinda** [comment]
@etoledom thanks for remarks ! We try to discuss `big font size` issue with Luke tomorrow

File: packages/components/src/mobile/bottom-sheet/cell.native.js
Comment (jbinda) [reply packages/components/src/mobile/bottom-sheet/cell.native.js]: fixed !

File: packages/components/src/mobile/bottom-sheet/range-cell.native.js
Comment (lukewalczak) [reply packages/components/src/mobile/bottom-sheet/range-cell.native.js]: * Reset button is placed within a `Cell` component, because we want to locate it next to the label (_Height in pixels_) which is defined there actually, otherwise it has to be positioned absolutely.

* Also we can block `allowReset` button option for other controls to avoid the case

**Comment by lukewalczak** [comment]
I did some accessibility tweaks (@etoledom @iamthomasbishop please look at it): 
  * made some space for `Reset` button, but then label is truncated
  * `TextInput` width is multiplied by `PixelRatio.getFontScale()` therefore we can display 4 digits
  * by default `TextInput` width is  `40`



default | medium | the largest
--- | --- | ---
def | medium | large



_pay no attention to the DONE button, something was wrong with a simulator_

File: packages/components/src/mobile/bottom-sheet/range-cell.native.js
Comment (lukewalczak) [inline packages/components/src/mobile/bottom-sheet/range-cell.native.js]: Honestly, I think that it's worth to save that info in redux store, because it can be useful in the future to control other components 🤔

Comment (etoledom) [reply packages/components/src/mobile/bottom-sheet/range-cell.native.js]: It does sound like it would be helpful! But maybe it's too much for this PR.

**Review by etoledom — COMMENTED** [review]
It looks so much better on bigger font sizes! Thanks for the update 🙏 

> made some space for Reset button, but then label is truncated

That's why is better to modify the layout to vertical when the text sizes are set to accessibility sizes. This is simple to do on iOS but not sure on RN. Anyway, I think it's good as it is for now 👍 

> TextInput width is multiplied by PixelRatio.getFontScale() therefore we can display 4 digits

Probably, having more than 4 digits is too much of an edge case for this context in particular. I wonder if there are other usages of the Range Control where the numbers could grow a lot bigger.

I think we can revisit if this case becomes a real problem.

From https://github.com/WordPress/gutenberg/pull/18157#discussion_r343600678
> Reset button is placed within a Cell component, because we want to locate it next to the label (Height in pixels) which is defined there actually, otherwise it has to be positioned absolutely.

Makes sense 👍 
I'm still not sure about having a specific `Reset` button that does nothing, and the reset logic has to be implemented on every instance of it. In this case I'd suggest to add it as a generic button (or "box" holder for any component).

But since it's the only component that uses it so far, it might be OK go forward with it and modify it if we need another control with Reset later on. I'm happy with what @pinarol decides.

**Comment by pinarol** [comment]
> otherwise it has to be positioned absolutely

> I'm still not sure about having a specific Reset button that does nothing, and the reset logic has to be implemented on every instance of it. In this case I'd suggest to add it as a generic button (or "box" holder for any component).

I agree both. I think we can make this better with a simple tweak. Let's not name it as Reset in BottomSheetCell if it doesn't reset anything inside that component. How about just naming it as "customAction" and injecting the title and event handler from outside? Something like this:

<BottomSheetCell
    ...
    customAction={title:_('Reset'), onAction:{this.onReset}}
/>

`"box" holder for any component` seems too upfront design to me at this point and without knowing the real use cases it can fail to meet future needs.

File: packages/components/src/mobile/bottom-sheet/cell.native.js
Comment (lukewalczak) [inline packages/components/src/mobile/bottom-sheet/cell.native.js]: @etoledom If we remove `numberOfLines={ 1 }` we can have a not truncated label and it will look like

Comment (lukewalczak) [reply packages/components/src/mobile/bottom-sheet/cell.native.js]: I've also checked how it will affect other settings:

before | after
--- | ---
Screenshot 2019-11-08 at 11 36 33 | Screenshot 2019-11-08 at 11 36 28

**Comment by lukewalczak** [comment]
> How about just naming it as "customAction" and injecting the title and event handler from outside?

Works for me! Going to adjust it :)

**Comment by etoledom** [comment]
> How about just naming it as "customAction" and injecting the title and event handler from outside?

Sounds good! That's what I meant by a `add it as a generic button` 👍

File: packages/components/src/mobile/bottom-sheet/cell.native.js
Comment (etoledom) [reply packages/components/src/mobile/bottom-sheet/cell.native.js]: Looks better than cropping! 
Let's go with it if it does't affect other instances of cells and controls.

**Comment by lukewalczak** [comment]
@etoledom @pinarol I've refactored it according to your suggestion and added:

`customActionButton={ { title: resetButtonText, handler: this.handleReset } }`

**Comment by pinarol** [comment]
Works for me! Also I did a quick testing and it is working pretty smoothly, i am very happy that we got rid of the jumpiness.

File: packages/components/src/mobile/bottom-sheet/cell.native.js
Comment (etoledom) [reply packages/components/src/mobile/bottom-sheet/cell.native.js]: > I've also checked how it will affect other settings

Again, better than cropping. Looks good to me :)

**Review by etoledom — APPROVED** [review]
Looks and work great 🎉 
Tested on iOS and Android.

Nice job @lukewalczak !

**Pull request merged**
